### PR TITLE
ARROW-9193: [C++] Avoid spurious intermediate string copy in ToDateHolder

### PR DESCRIPTION
**Pull request by projjal**

### CODE REVIEW

**Comment by projjal** [comment]
Hi @wesm can you take a look at this. It seems that the ParseTimestampStrptime function was recently changed from taking c-style null terminated string input to c++ style. If the input string was already null terminated then this method significantly slows down due to extra allocation and copies from constructor and c_str(). From the benchmark test that I have added it seems that it gets about 2-2.5x slower. Would it make sense to add another method that explicitly takes null-terminated string as input to avoid these copies.

**Comment by wesm** [comment]
Seems reasonable, I or @pitrou will have to take a closer look. Can we centralize the benchmarks, though, so we don't have benchmarks for stuff in src/arrow/util in src/gandiva?

File: cpp/src/arrow/util/value_parsing.h [hunk 1]
```diff
@@ -565,6 +565,39 @@ static inline bool ParseTimestampStrptime(const char* buf, size_t length,
   return true;
 }
 
+/// \brief Returns time since the UNIX epoch in the requested unit. Takes null terminated
+/// buffer as argument
+static inline bool ParseTimestampStrptimeFromCString(const char* buf, size_t length,
```
Comment (pitrou) [inline cpp/src/arrow/util/value_parsing.h:570]: Can you please refactor these two functions to avoid duplicated code?

**Review by pitrou — CHANGES_REQUESTED** [review]

**Comment by pitrou** [comment]
@projjal Where do the null-terminated strings come from? Doesn't Gandiva operate on Arrow data?

File: cpp/src/gandiva/to_date_holder.cc [hunk 2]
```diff
@@ -83,7 +83,7 @@ int64_t ToDateHolder::operator()(ExecutionContext* context, const std::string& d
   // 1. processes date that do not match the format.
   // 2. does not process time in format +08:00 (or) id.
   int64_t seconds_since_epoch = 0;
-  if (!::arrow::internal::ParseTimestampStrptime(
+  if (!::arrow::internal::ParseTimestampStrptimeFromCString(
           data.c_str(), data.length(), pattern_.c_str(),
```
Comment (projjal) [inline cpp/src/gandiva/to_date_holder.cc:87]: @pitrou here data.c_str() returns null-terminated string which we pass to parsetimestamp method

Comment (pitrou) [reply cpp/src/gandiva/to_date_holder.cc:87]: Understood. But where does that data come from? Why do you have a `std::string`?

Comment (projjal) [reply cpp/src/gandiva/to_date_holder.cc:87]: It is created here : https://github.com/apache/arrow/blob/295e2c11642f1b32aff1473de3f89ee02baada8a/cpp/src/gandiva/gdv_function_stubs.cc#L60

Comment (pitrou) [reply cpp/src/gandiva/to_date_holder.cc:87]: Ah. Well, if you're concerned about performance, you shouldn't create a `std::string` there.

Comment (projjal) [reply cpp/src/gandiva/to_date_holder.cc:87]: Makes sense. I will update the patch to pass the data buffer directly to the orginal parsetimestamp method. The new method won't be necessary then i guess.

**Review by pitrou — APPROVED** [review]
+1, thank you

**Comment by pitrou** [comment]
@projjal Can you fix the Windows build failures (see AppVeyor build):
```
C:/projects/arrow/cpp/src/gandiva/to_date_holder_test.cc(55): error C2220: warning treated as error - no 'object' file generated
C:/projects/arrow/cpp/src/gandiva/to_date_holder_test.cc(55): warning C4267: 'argument': conversion from 'size_t' to 'int', possible loss of data
C:/projects/arrow/cpp/src/gandiva/to_date_holder_test.cc(60): warning C4267: 'argument': conversion from 'size_t' to 'int', possible loss of data
C:/projects/arrow/cpp/src/gandiva/to_date_holder_test.cc(65): warning C4267: 'argument': conversion from 'size_t' to 'int', possible loss of data
C:/projects/arrow/cpp/src/gandiva/to_date_holder_test.cc(79): warning C4267: 'argument': conversion from 'size_t' to 'int', possible loss of data
C:/projects/arrow/cpp/src/gandiva/to_date_holder_test.cc(84): warning C4267: 'argument': conversion from 'size_t' to 'int', possible loss of data
C:/projects/arrow/cpp/src/gandiva/to_date_holder_test.cc(96): warning C4267: 'argument': conversion from 'size_t' to 'int', possible loss of data
C:/projects/arrow/cpp/src/gandiva/to_date_holder_test.cc(101): warning C4267: 'argument': conversion from 'size_t' to 'int', possible loss of data
C:/projects/arrow/cpp/src/gandiva/to_date_holder_test.cc(106): warning C4267: 'argument': conversion from 'size_t' to 'int', possible loss of data
C:/projects/arrow/cpp/src/gandiva/to_date_holder_test.cc(111): warning C4267: 'argument': conversion from 'size_t' to 'int', possible loss of data
C:/projects/arrow/cpp/src/gandiva/to_date_holder_test.cc(117): warning C4267: 'argument': conversion from 'size_t' to 'int', possible loss of data
C:/projects/arrow/cpp/src/gandiva/to_date_holder_test.cc(131): warning C4267: 'argument': conversion from 'size_t' to 'int', possible loss of data
```
?

**Comment by fsaintjacques** [comment]
Any reason not to use a `util::string_view`?

**Comment by projjal** [comment]
> Any reason not to use a `util::string_view`?

Since the toDate function here takes a char array and just passes it onto the parseTimestamp method, using a util::string_view is probably not required.

**Pull request closed without merging**